README.md

# Mezzanine Facebook Feed

Connect Mezzanine sites to Facebook feeds

## Installation

1. Install via pip: `pip install mezzanine-facebook-feed`.
2. Add `fbfeed` to your `INSTALLED_APPS`.
3. Run migrations.

## Usage

### Add Facebook App credentials

Add your Facebook App ID and Secret to Mezzanine's setting page. Configure your App with a redirect URI of `<your domain>/admin/fbfeed/access/redirect/`.

### Create an Access instance

To access Facebook you need to create access tokens so Mezzanine can run queries on your behalf. To create access tokens visit your admin site and click `FB Feed > Access`.

1. Create an Access instance and give it a name (this is arbitrary and for your use only).
1. Set Facebook ID to the numeric ID of the user or page you want to query. If you want to query your own data, enter "me".
1. Save the Access instance and open it again in the admin.
1. Click the "Get Access Token" button in the upper-right corner.
1. You will be redirected to Facebook. Complete the login process.
1. If everything goes well, you'll see the access token in the Access admin page.
1. Otherwise an error message will be shown on the top of the admin site.

### Display your data in templates

```django
{% load fbfeed_tags %}
{% fbfeed_photos as photos %}
{% for photo in photos %}
  <a href="{{ photo.link }}">
    <img src="{{ photo.images.0.source }}" alt="{{ photo.alt_text }}">
  </a>
{% endfor %}
```

## Contributing

Review contribution guidelines at [CONTRIBUTING.md].

[CONTRIBUTING.md]: CONTRIBUTING.md


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.3.1](https://github.com/unplugstudio/mezzanine-facebook-feed/compare/v0.3.0...v0.3.1) (2020-03-06)


### Features

* add configurable Facebook Scopes ([1993d54](https://github.com/unplugstudio/mezzanine-facebook-feed/commit/1993d5480f2526296792bc5fb1cbd6eca8777f23))
* show error messages on the redirect endpoint ([25b008e](https://github.com/unplugstudio/mezzanine-facebook-feed/commit/25b008ed0da64ae00437e33d742c15c805587b9f))


### Bug Fixes

* fix unicode errors in Python 2 ([07aedde](https://github.com/unplugstudio/mezzanine-facebook-feed/commit/07aedde0f6afba9d590844613b97140c7987cc01))

# [0.3.0](https://github.com/unplugstudio/mezzanine-facebook-feed/compare/v0.2.0...v0.3.0) (2020-02-21)


### Features

* add logging to fbfeed_photos ([afa2c16](https://github.com/unplugstudio/mezzanine-facebook-feed/commit/afa2c16))



# [0.2.0](https://github.com/unplugstudio/mezzanine-facebook-feed/compare/1f60058...v0.2.0) (2020-02-21)


### Features

* connect to Facebook and fetch photos ([1f60058](https://github.com/unplugstudio/mezzanine-facebook-feed/commit/1f60058))





fbfeed/admin.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals, absolute_import

from builtins import str

from django.conf.urls import url
from django.contrib import admin
from django.shortcuts import redirect, get_object_or_404

from mezzanine.utils.urls import admin_url

from .models import Access, FacebookSettingError, GraphAPIError


@admin.register(Access)
class AccessAdmin(admin.ModelAdmin):
    readonly_fields = ["token", "token_type", "expires"]

    def change_view(self, request, object_id, form_url="", extra_context=None):
        """
        Don't render the change view until the Facebook settings are set
        """
        access = get_object_or_404(Access, id=object_id)
        try:
            access.app_id
            access.app_secret
        except FacebookSettingError as e:
            self.message_user(request, str(e))
            return redirect("admin:conf_setting_changelist")
        return super(AccessAdmin, self).change_view(
            request, object_id, form_url=form_url, extra_context=extra_context
        )

    def redirect_endpoint(self, request):
        """
        Process the request after the user has logged into Facebook
        """
        code = request.GET.get("code")
        access = get_object_or_404(Access, id=request.GET.get("state", 0))
        try:
            access.get_token_from_code(code)
        except (GraphAPIError, FacebookSettingError) as e:
            self.message_user(request, str(e))
        else:
            self.message_user(request, "Access token updated")

        return redirect(admin_url(Access, "change", access.id))

    def get_urls(self):
        urls = [
            url(
                r"^redirect/$",
                self.admin_site.admin_view(self.redirect_endpoint),
                name="fbfeed_redirect_endpoint",
            )
        ]
        return urls + super(AccessAdmin, self).get_urls()


fbfeed/migrations/0001_initial.py
# -*- coding: utf-8 -*-
# Generated by Django 1.11.20 on 2020-02-21 20:53
from __future__ import unicode_literals

from django.db import migrations, models


class Migration(migrations.Migration):

    initial = True

    dependencies = [
    ]

    operations = [
        migrations.CreateModel(
            name='Access',
            fields=[
                ('id', models.AutoField(auto_created=True, primary_key=True, serialize=False, verbose_name='ID')),
                ('created', models.DateTimeField(editable=False, null=True)),
                ('updated', models.DateTimeField(editable=False, null=True)),
                ('name', models.CharField(max_length=200, verbose_name='Name')),
                ('facebook_id', models.CharField(default='me', help_text="Use 'me' to view your own posts", max_length=100, verbose_name='Facebook ID')),
                ('token', models.CharField(blank=True, max_length=500, verbose_name='Access token')),
                ('token_type', models.CharField(blank=True, max_length=200, verbose_name='Token type')),
                ('expires', models.DateTimeField(blank=True, null=True, verbose_name='Expires')),
            ],
            options={
                'verbose_name': 'access',
                'verbose_name_plural': 'accesses',
            },
        ),
    ]


fbfeed/models.py
# -*- coding: utf-8 -*-
from __future__ import unicode_literals, absolute_import

try:  # Python 3
    from urllib.parse import urlencode
except ImportError:  # Puthon 2
    from urllib import urlencode

import requests

from datetime import timedelta

from django.core.cache import cache
from django.core.urlresolvers import reverse
from django.db import models
from django.utils.encoding import python_2_unicode_compatible
from django.utils.timezone import now

from mezzanine.conf import settings
from mezzanine.core.models import TimeStamped
from mezzanine.utils.sites import current_request


class FacebookSettingError(Exception):
    pass


class GraphAPIError(Exception):
    """
    https://github.com/mobolic/facebook-sdk/blob/master/facebook/__init__.py
    """

    def __init__(self, result):
        self.result = result
        self.code = None
        self.error_subcode = None

        try:
            self.type = result["error_code"]
        except (KeyError, TypeError):
            self.type = ""

        # OAuth 2.0 Draft 10
        try:
            self.message = result["error_description"]
        except (KeyError, TypeError):
            # OAuth 2.0 Draft 00
            try:
                self.message = result["error"]["message"]
                self.code = result["error"].get("code")
                self.error_subcode = result["error"].get("error_subcode")
                if not self.type:
                    self.type = result["error"].get("type", "")
            except (KeyError, TypeError):
                # REST server style
                try:
                    self.message = result["error_msg"]
                except (KeyError, TypeError):
                    self.message = result

        Exception.__init__(self, self.message)


@python_2_unicode_compatible
class Access(TimeStamped):

    name = models.CharField("Name", max_length=200)
    facebook_id = models.CharField(
        "Facebook ID",
        max_length=100,
        default="me",
        help_text="Use 'me' to view your own posts",
    )
    token = models.CharField("Access token", max_length=500, blank=True)
    token_type = models.CharField("Token type", max_length=200, blank=True)
    expires = models.DateTimeField("Expires", blank=True, null=True)

    class Meta:
        verbose_name = "access"
        verbose_name_plural = "accesses"

    def __str__(self):
        return self.name

    @property
    def app_id(self):
        value = settings.FACEBOOK_APP_ID
        if not value:
            raise FacebookSettingError("Set the Facebook App ID in settings")
        return value

    @property
    def app_secret(self):
        value = settings.FACEBOOK_APP_SECRET
        if not value:
            raise FacebookSettingError("Set the Facebook App Secret in settings")
        return value

    def get_redirect_url(self):
        """
        Build the redirect URL required by Facebook (including the domain and scheme)
        """
        request = current_request()
        return request.build_absolute_uri(reverse("admin:fbfeed_redirect_endpoint"))

    def get_login_url(self):
        """
        Build the URL that invokes the Facebook login dialog
        """
        return "https://www.facebook.com/v6.0/dialog/oauth?" + urlencode(
            {
                "client_id": self.app_id,
                "redirect_uri": self.get_redirect_url(),
                "state": self.id,
            }
        )

    def get_token_from_code(self, code):
        """
        Exchange the short code generated by the Facebook login dialog for an access token
        """
        url = "https://graph.facebook.com/v6.0/oauth/access_token?" + urlencode(
            {
                "client_id": self.app_id,
                "redirect_uri": self.get_redirect_url(),
                "client_secret": self.app_secret,
                "code": code,
            }
        )
        response = requests.get(url).json()
        if "error" in response:
            raise GraphAPIError(response)
        self.token = response["access_token"]
        self.token_type = response["token_type"]
        self.expires = now() + timedelta(seconds=response["expires_in"])
        self.save()

    def get_photos(self, limit=10, clear_cache=False):
        """
        Get photos as a list of Photo nodes.
        https://developers.facebook.com/docs/graph-api/reference/photo/
        """
        CACHE_KEY = "fbfeed_photos_{}".format(self.pk)
        data = cache.get(CACHE_KEY)

        # Serve from cache if available
        if not clear_cache and data is not None:
            return data[:limit]

        url = "https://graph.facebook.com/v6.0/{}/photos?".format(self.facebook_id)
        url += urlencode(
            {
                "fields": "alt_text,link,images",
                "type": "uploaded",
                "limit": limit,
                "access_token": self.token,
            }
        )
        response = requests.get(url).json()

        if "error" in response:
            raise GraphAPIError(response)

        data = response.get("data", [])
        cache.set(CACHE_KEY, data, 60 * 15)
        return data[:limit]
